### PR TITLE
#187419022 Implement Cart Management Functionality

### PR DESCRIPTION
**Pull request by Hakizimana-Clement**
#### What does this PR do?
This PR introduces cart management functionality, allowing buyers to add items to their cart and view the cart.

#### Description of Task to be completed:
- Implement the UI components for adding items to the cart (quantity input and "Add to Cart" button).
- Create a confirmation message that displays the added item's name and quantity.
- Ensure that the form for selecting quantity allows only positive integers.
#### How should this be manually tested?
1. Clone the repo and cd into it
2. Checkout to ft-cart-management-187419022 and run npm install
3. Run npm run dev to start the application locally
4. Navigate to a product page.
5. Select a quantity and click the "Add to Cart" button.
6. Verify that the item is added to the cart and that the confirmation message appears.
7. Test the "View Cart" feature by clicking the corresponding button
#### Pivot tracker Id
#187419022

### CODE REVIEW

**Review by cyusasnave — COMMENTED** [review]
About the cart everything is working well as expected, though you're making too much request when increasing or descreasing the quantity. Maybe @leandreAlly can say something about this.
Screenshot 2024-07-08 at 09 26 24
You can see how the he is adding to cart and fetching the user cart again.

File: test/components/buttons/CartQuantity.test.tsx [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+// 	// test('should show error message if dispatch fails', async () => {
+// 	// 	dispatchMock.mockRejectedValueOnce(new Error('Test error'));
+// 	// 	customRender(<CartQuantity productId="1" />);
+// 	// 	const incrementButton = screen.getByText('+');
+
+// 	// 	await fireEvent.click(incrementButton);
+
+// 	// 	expect(showErrorMessageMock).toHaveBeenCalledWith('Test error');
+// 	// });
```
Comment (Tuyisenge2) [inline test/components/buttons/CartQuantity.test.tsx]: remove this code if they are not used anymore

**Review by Tuyisenge2 — APPROVED** [review]

File: test/components/buttons/CartQuantity.test.tsx (hunk 1 above)
Comment (Hakizimana-Clement) [reply test/components/buttons/CartQuantity.test.tsx]: i remove it @Tuyisenge2

**Comment by Hakizimana-Clement** [comment]
> About the cart everything is working well as expected, though you're making too much request when increasing or descreasing the quantity. Maybe @leandreAlly can say something about this. Screenshot 2024-07-08 at 09 26 24 You can see how the he is adding to cart and fetching the user cart again.



> About the cart everything is working well as expected, though you're making too much request when increasing or descreasing the quantity. Maybe @leandreAlly can say something about this. Screenshot 2024-07-08 at 09 26 24 You can see how the he is adding to cart and fetching the user cart again.

File: src/components/buttons/RemoveCart.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { useAppDispatch } from '../../redux/hooks/hooks';
+import IconLoader from '../Loaders/IconLoader';
+
+interface RemoveCartProps {
+	productId: string;
+}
+
+const RemoveCartButton = ({ productId }: RemoveCartProps) => {
+	const { showErrorMessage, showSuccessMessage } = useToast();
+	const [isButtonLoading, setIsButtonLoading] = useState(false);
```
Comment (GarrixA) [inline src/components/buttons/RemoveCart.tsx]: Remove IsLoading and use redux states

File: src/components/buttons/RemoveCart.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	const dispatch = useAppDispatch();
+
+	const handleClick = async () => {
+		setIsButtonLoading(true);
```
Comment (GarrixA) [inline src/components/buttons/RemoveCart.tsx]: Remove this line

File: src/components/buttons/RemoveCart.tsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+			showSuccessMessage('Cart successfully removed');
+		} catch (e) {
+			const err = e as DynamicData;
+			showErrorMessage(
+				err?.data?.message ||
+					err?.message ||
+					'Unknown error occured! Please try again!',
+			);
+		} finally {
+			setIsButtonLoading(false);
```
Comment (GarrixA) [inline src/components/buttons/RemoveCart.tsx]: Remove this too

File: src/redux/features/cartSlice.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		}
+	},
+);
+
+export const cartQuantities = createAsyncThunk(
+	'cartQuantity',
+	async ({ productId, quantity }: cartDataToSend, { rejectWithValue }) => {
+		try {
+			await API.post('/carts', { productId, quantity });
+			const updatedCartData = await API.get('/carts'); // Fetch updated cart data
```
Comment (GarrixA) [inline src/redux/features/cartSlice.ts]: Remove this comment

File: src/redux/features/cartSlice.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+			(state, action: PayloadAction<DynamicData>) => {
+				state.isLoading = false;
+				state.carts = action.payload.data;
+				state.numberOfItem = action.payload.data.products.length;
+			},
+		);
+
+		builder.addCase(
+			removeCarts.rejected,
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (GarrixA) [inline src/redux/features/cartSlice.ts]: Instead of duplicating this, you can use it on the top of the file to handle all the cases

Comment (Hakizimana-Clement) [reply src/redux/features/cartSlice.ts]: yes, let me fix

File: src/redux/features/cartSlice.ts (hunk 5 above)
Comment (Hakizimana-Clement) [reply src/redux/features/cartSlice.ts]: yes, i remove it.

File: src/components/buttons/RemoveCart.tsx (hunk 2 above)
Comment (Hakizimana-Clement) [reply src/components/buttons/RemoveCart.tsx]: yes, i correct it

File: src/components/buttons/RemoveCart.tsx (hunk 4 above)
Comment (Hakizimana-Clement) [reply src/components/buttons/RemoveCart.tsx]: yes, let me remove it

File: src/components/buttons/RemoveCart.tsx (hunk 3 above)
Comment (Hakizimana-Clement) [reply src/components/buttons/RemoveCart.tsx]: yes

**Comment by leandreAlly** [comment]
> About the cart everything is working well as expected, though you're making too much request when increasing or descreasing the quantity. Maybe @leandreAlly can say something about this. Screenshot 2024-07-08 at 09 26 24 You can see how the he is adding to cart and fetching the user cart again.

@cyusasnave thanks for bringing this out, we will check in together and see if we can come up with efficient solution.

File: test/pages/SingleProduct.test.tsx [hunk 7]
```diff
@@ -1,3 +1,162 @@
+// import { render, screen } from '@testing-library/react';
```
Comment (cyusasnave) [inline test/pages/SingleProduct.test.tsx]: I think you don't need this commented lines, you can remove them. Then we're good on this PR

**Review by cyusasnave — CHANGES_REQUESTED** [review]

**Review by cyusasnave — APPROVED** [review]

**Review by GarrixA — APPROVED** [review]

**Review by leandreAlly — APPROVED** [review]

**Pull request merged**